src/main/java/com/snowflake/kafka/connector/internal/streaming/StreamingClientProvider.java

```diff
@@ -47,72 +52,137 @@ public static StreamingClientProvider getStreamingClientProviderInstance() {
+
     if (Boolean.parseBoolean(
         connectorConfig.getOrDefault(
             SnowflakeSinkConnectorConfig.ENABLE_STREAMING_CLIENT_OPTIMIZATION_CONFIG,
             Boolean.toString(ENABLE_STREAMING_CLIENT_OPTIMIZATION_DEFAULT)))) {
       LOGGER.info(
-          "Streaming client optimization is enabled, returning the existing streaming client if"
-              + " valid");
-      this.providerLock.lock();
-      // recreate streaming client if needed
-      if (!StreamingClientHandler.isClientValid(this.parameterEnabledClient)) {
-        LOGGER.error("Current streaming client is invalid, recreating client");
-        this.parameterEnabledClient = this.streamingClientHandler.createClient(connectorConfig);
+          "Streaming client optimization is enabled per worker node. Reusing valid clients when"
+              + " possible");
+      resultClient = this.registeredClients.get(clientProperties);
+
+      // refresh if registered client is invalid
+      if (!StreamingClientHandler.isClientValid(resultClient)) {
+        resultClient = this.streamingClientHandler.createClient(clientProperties);
+        this.registeredClients.put(clientProperties, resultClient);
       }
-      this.providerLock.unlock();
-      return this.parameterEnabledClient;
     } else {
-      LOGGER.info("Streaming client optimization is disabled, creating a new streaming client");
-      return this.streamingClientHandler.createClient(connectorConfig);
+      resultClient = this.streamingClientHandler.createClient(clientProperties);
+      LOGGER.info(
+          "Streaming client optimization is disabled, creating a new streaming client with name:"
+              + " {}",
+          resultClient.getName());
     }
+
+    return resultClient;
   }
 
   /**
-   * Closes the given client
+   * Closes the given client and deregisters it from the cache if necessary. It will also call close
+   * on the registered client if exists, which should be the same as the given client so the call
+   * will no-op.
    *
+   * @param connectorConfig The configuration to deregister from the cache
    * @param client The client to be closed
    */
-  public void closeClient(SnowflakeStreamingIngestClient client) {
-    this.providerLock.lock();
+  public void closeClient(
+      Map<String, String> connectorConfig, SnowflakeStreamingIngestClient client) {
+    StreamingClientProperties clientProperties = new StreamingClientProperties(connectorConfig);
+
+    // invalidate cache
+    SnowflakeStreamingIngestClient registeredClient =
+        this.registeredClients.getIfPresent(clientProperties);
+    if (registeredClient != null) {
+      // invalidations are processed on the next get or in the background, so we still need to close
+      // the client here
+      this.registeredClients.invalidate(clientProperties);
+      this.streamingClientHandler.closeClient(registeredClient);
```

Comment:
duplicate as line 162?

Reply:
Also do you need to check whether the client is still valid before closing it?

Reply:
We call close on the given client and the client registered in the cache. Technically they should be the same client, however I prefer to call close on both just in case the given client is different or somehow the registered cache was corrupted

The streamingClientHandler will check if the client is valid before calling close, so the extra close call will no-op if it is invalid